packages/contract/src/pay/PayIntent.sol

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.12;
+
+import "openzeppelin-contracts/contracts/proxy/utils/Initializable.sol";
+import "openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import "./TokenRefund.sol";
+import "./TokenUtils.sol";
+import "../interfaces/IDaimoPayBridger.sol";
+
+/// @dev Represents an intended call: "make X of token Y show up on chain Z, then
+///      use it to do an arbitrary contract call".
+struct PayIntent {
+    /// @dev Intent only executes on given target chain.
+    uint256 toChainId;
+    /// @dev Expected token after bridging to the destination chain.
+    TokenAmount bridgeTokenOut;
+    /// @dev Expected token amount after swapping on the destination chain.
+    TokenAmount finalCallToken;
+    /// @dev Destination on target chain. If dest.data != "" specifies a call,
+    ///     (token, amount) is approved. Otherwise, it's transferred to dest.to
+    Call finalCall;
+    /// @dev Escrow contract for fast-finish. Will typically be the DaimoPay
+    ///      contract.
+    address payable escrow;
+    /// @dev Address to refund tokens if call fails, or zero.
+    address refundAddress;
+    /// @dev Nonce. PayIntent receiving addresses are one-time use.
+    uint256 nonce;
+}
+
+/// @dev Calculates the intent hash of a PayIntent struct
+/// @param intent The PayIntent struct to hash
+/// @return The keccak256 hash of the encoded PayIntent
+function calcIntentHash(PayIntent calldata intent) pure returns (bytes32) {
+    return keccak256(abi.encode(intent));
+}
+
+/// @dev This is an ephemeral intent contract. Any supported tokens sent to this
+///      address on any supported chain are forwarded, via a combination of
+///      bridging and swapping, into a specified call on a destination chain.
+contract PayIntentContract is Initializable {
+    using SafeERC20 for IERC20;
+
+    /// @dev Save gas by minimizing storage to a single word. This makes intents
+    ///      usable on L1. intentHash = keccak(abi.encode(PayIntent))
+    bytes32 intentHash;
+
+    /// @dev Runs at deploy time. Singleton implementation contract = no init,
+    ///      no state. All other methods are called via proxy.
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(bytes32 _intentHash) public initializer {
+        intentHash = _intentHash;
+    }
+
+    /// Called on the source chain to initiate the intent. Sends funds to dest
+    /// chain.
+    function sendAndSelfDestruct(
+        PayIntent calldata intent,
+        IDaimoPayBridger bridger,
```

Comment:
i think the bridger has to be trusted? = either committed to in the intent address, or a constant in this file

otherwise, someone can pass in a fund-stealing bridger, right?

Reply:
the bridger is currently stored as immutable in the main `DaimoPay.sol` contract and gets passed as an arg in `sendAndSelfDestruct`. since the intent specifies the `DaimoPay` contract as the escrow, `DaimoPay` is the only contract that can deploy code to our intent addrs, so there's no way to pass in a malicious bridger

- i considered hardcoding the bridger in `PayIntent.sol`, but that makes deployment a little tricky
- also considered passing the bridger into the constructor of `PayIntentFactory`, but then `PayIntent` would need to store the `bridger` in it's storage